pkg/pillar/cmd/zedrouter/ipam.go

```diff
@@ -17,27 +17,33 @@ func (z *zedrouter) generateBridgeMAC(brNum int) net.HardwareAddr {
 	return net.HardwareAddr{0x00, 0x16, 0x3e, 0x06, 0x00, byte(brNum)}
 }
 
-// generateAppMac picks a fixed address for Local and uses a fixed hash for Switch
-// which still produces a stable MAC address for a given app instance.
+// generateAppMac calculates random but stable (not changing across reboots) MAC address
+// for a given app instance.
+// Even for local network instances we generate random MAC address that with a high
+// probability will not collide with other apps on other edge nodes. From networking
+// point of view this is not necessary, but there are apps that use MAC address as some
+// sort of ID and could fail to function properly if there is a collision across edge
+// nodes.
+// We use locally administered MAC addresses as opposed to universally administered
+// (UAA) - where 3 octets are used as OUI. The reason is that with only 24 bits
+// remaining and available for randomness, the chances of at least one collision
+// are fairly high (e.g. 3% for 1000 app interfaces).
 func (z *zedrouter) generateAppMac(appUUID uuid.UUID, ulNum int, appNum int,
 	netInstStatus *types.NetworkInstanceStatus) net.HardwareAddr {
-	switch netInstStatus.Type {
-	case types.NetworkInstanceTypeSwitch:
-		h := sha256.New()
-		h.Write(appUUID[:])
-		h.Write(netInstStatus.UUIDandVersion.UUID[:])
-		nums := make([]byte, 2)
-		nums[0] = byte(ulNum)
-		nums[1] = byte(appNum)
-		h.Write(nums)
-		hash := h.Sum(nil)
-		return net.HardwareAddr{0x02, 0x16, 0x3e, hash[0], hash[1], hash[2]}
-
-	case types.NetworkInstanceTypeLocal:
-		// Room to handle multiple underlays in 5th byte
-		return net.HardwareAddr{0x00, 0x16, 0x3e, 0x00, byte(ulNum), byte(appNum)}
-	}
-	return nil
+	h := sha256.New()
+	h.Write(appUUID[:])
+	h.Write(netInstStatus.UUIDandVersion.UUID[:])
+	nums := make([]byte, 2)
+	nums[0] = byte(ulNum)
+	nums[1] = byte(appNum)
+	h.Write(nums)
+	hash := h.Sum(nil)
+	mac := net.HardwareAddr{hash[0], hash[1], hash[2], hash[3], hash[4], hash[5]}
```

Comment:
Two issues:
1. This changes the MAC addresses for existing app instances deployed using switch network instances, and that is likely to get a different IP address from the external DHCP server. That seems like a risky and uncecessary change since we don't know what the impact will be on existing deployments.
2. For local network instances I think it is OK to have 46 bits of randomness since these MAC addresses do not appear on an external Ethernet network. However, if you read the specifications from the IEEE 802 RAC they are very clear that it is the owner of the OUI who also owns the corresponding MAC addresses witn the local bit set. I think we should document this in the code.

Reply:
1. True, I reverted the behavior for generating MACs for switch network instances back to what it was before.
2. I have updated the function comment - let me know if you think we should add more clarification and maybe include something from IEEE specs.